File: services/agent-api/app/db/models/presence.py
"""PresenceOpportunity model."""

import uuid
from datetime import datetime

from sqlalchemy import ARRAY, Boolean, DateTime, Double, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import UUIDMixin, TimestampMixin, MetadataMixin, Base


class PresenceOpportunity(Base, UUIDMixin, TimestampMixin, MetadataMixin):
    __tablename__ = "presence_opportunities"

    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    companion_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("companions.id"), nullable=False
    )
    conversation_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("conversations.id"), nullable=True
    )
    co_presence_session_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("co_presence_sessions.id"), nullable=True
    )

    type: Mapped[str] = mapped_column(String(50), nullable=False)
    title: Mapped[str] = mapped_column(Text, nullable=False)
    message: Mapped[str | None] = mapped_column(Text, nullable=True)
    reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    evidence_memory_ids: Mapped[list[uuid.UUID]] = mapped_column(
        ARRAY(UUID(as_uuid=True)), default=list, server_default="{}"
    )
    evidence_message_ids: Mapped[list[uuid.UUID]] = mapped_column(
        ARRAY(UUID(as_uuid=True)), default=list, server_default="{}"
    )
    evidence_growth_ids: Mapped[list[uuid.UUID]] = mapped_column(
        ARRAY(UUID(as_uuid=True)), default=list, server_default="{}"
    )

    priority: Mapped[float] = mapped_column(Double, default=0.0)
    urgency: Mapped[float] = mapped_column(Double, default=0.0)
    sensitivity: Mapped[float] = mapped_column(Double, default=0.0)
    interruption_risk: Mapped[float] = mapped_column(Double, default=0.0)
    recommended_surface: Mapped[str] = mapped_column(String(50), default="queue")

    status: Mapped[str] = mapped_column(String(50), default="queued")
    snoozed_until: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    expires_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    user_action: Mapped[str | None] = mapped_column(String(50), nullable=True)
    reward: Mapped[float | None] = mapped_column(Double, nullable=True)
    score_json: Mapped[dict] = mapped_column(JSONB, default=dict, server_default="{}")

    # ── Continuity: feedback / timing / calibration ────────────────────
    dismissed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    dismissed_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    accepted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    suppress_type_rule_applied: Mapped[bool] = mapped_column(Boolean, default=False)
    feedback_event_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("feedback_events.id"), nullable=True
    )
    feedback_label: Mapped[str | None] = mapped_column(String, nullable=True)
    timing_score: Mapped[float] = mapped_column(Double, default=0.5)
    type_affinity_snapshot: Mapped[float] = mapped_column(Double, default=0.5)
    opportunity_context_hash: Mapped[str | None] = mapped_column(Text, nullable=True)
    meaningful_silence_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    calibration_json: Mapped[dict] = mapped_column(JSONB, default=dict, server_default="{}")

File: apps/web/lib/api/presence.ts
import { api } from "./client";

export interface PresenceOpportunity {
  id: string; type?: string; title?: string; status: string;
  reason?: string; summary?: string; message?: string;
  priority?: number; urgency?: number; sensitivity?: number;
  interruption_risk?: number; recommended_surface?: string;
  // Extended product fields
  snoozed_until?: string | null; dismissed_at?: string | null;
  dismissed_reason?: string | null; accepted_at?: string | null;
  suppress_type_rule_applied?: boolean;
  timing_score?: number; type_affinity_snapshot?: number;
  meaningful_silence_reason?: string | null;
  feedback_label?: string | null; feedback_event_id?: string;
  calibration_json?: Record<string, unknown>;
  opportunity_context_hash?: string;
  created_at?: string;
}

export function listOpportunities(params?: Record<string,string>) {
  const qs = params ? "?" + new URLSearchParams(params).toString() : "";
  return api.get<{ items: PresenceOpportunity[]; total: number }>(`/presence/opportunities${qs}`);
}
export function acceptOpportunity(id: string) { return api.post(`/presence/opportunities/${id}/accept`); }
export function dismissOpportunity(id: string) { return api.post(`/presence/opportunities/${id}/dismiss`); }
export function snoozeOpportunity(id: string) { return api.post(`/presence/opportunities/${id}/snooze`); }
export function suppressOpportunityType(id: string) { return api.post(`/presence/opportunities/${id}/suppress-type`); }

export type PresenceDestinationMode = "bound_conversation" | "new_conversation_per_delivery";
export interface PresenceSchedule {
  id: string; user_id: string; companion_id: string; status: "active" | "paused";
  pause_reason?: string | null; destination_mode: PresenceDestinationMode;
  bound_conversation_id?: string | null; latest_created_conversation_id?: string | null;
  timezone: string; weekdays: number[]; timing_mode: "fixed" | "random_window";
  fixed_minute_of_day: number; window_start_minute: number; window_end_minute: number;
  cadence_mode: "fixed" | "random_interval"; fixed_interval_minutes: number;
  random_interval_min_minutes: number; random_interval_max_minutes: number;
  revision: number; next_occurrence_at?: string | null; last_delivered_at?: string | null; updated_at: string;
}
export type PresenceScheduleInput = Omit<PresenceSchedule, "id" | "user_id" | "companion_id" | "pause_reason" | "latest_created_conversation_id" | "next_occurrence_at" | "last_delivered_at" | "updated_at" | "revision"> & { expected_revision?: number };
export interface PresenceOccurrence {
  id: string; status: string; sequence_no: number; scheduled_for: string; next_attempt_at?: string | null; delivered_at?: string | null;
  attempt_count: number; conversation_id?: string | null; message_id?: string | null;
  suppression_reason?: string | null; error_code?: string | null;
  random_draw: Record<string, unknown>; delivery_evidence: Record<string, unknown>;
}
export interface PresenceConfigurationVersions {
  schedule_revision: number | null;
  policy_updated_at: string | null;
  persona_updated_at: string;
  boundary_updated_at: string;
}
export interface PresenceConfigurationValue {
  enabled: boolean;
  proactive_level: "low" | "medium" | "high";
  presence_style: "quiet" | "balanced" | "expressive";
  notification_surface: "hub_queue_only" | "allow_light_notification" | "disabled";
  meaningful_silence_enabled: boolean;
  quiet_hours: { enabled: boolean; start: string; end: string };
  max_presence_per_day: number;
  destination_mode: PresenceDestinationMode;
  bound_conversation_id: string | null;
  timezone: string;
  weekdays: number[];
  timing_mode: "fixed" | "random_window";
  fixed_minute_of_day: number;
  window_start_minute: number;
  window_end_minute: number;
  cadence_mode: "fixed" | "random_interval";
  fixed_interval_minutes: number;
  random_interval_min_minutes: number;
  random_interval_max_minutes: number;
}
export interface PresenceConfiguration {
  contract_version: "presence-configuration.v1";
  user_id: string;
  companion_id: string;
  versions: PresenceConfigurationVersions;
  configuration: PresenceConfigurationValue;
  runtime: { next_occurrence_at: string | null; last_delivered_at: string | null };
  consistency: {
    status: "aligned" | "needs_save_to_align";
    warnings: string[];
    canonical_quiet_hours_source: "boundary_settings";
    derived_profile_projection: boolean;
    derived_min_interval_from_schedule: boolean;
  };
}
export type PresenceConfigurationInput = PresenceConfigurationValue & {
  expected_schedule_revision: number | null;
  expected_policy_updated_at: string | null;
  expected_persona_updated_at: string;
  expected_boundary_updated_at: string;
};
export function getPresenceConfiguration(companionId: string, userId: string) {
  return api.get<PresenceConfiguration>(`/companions/${companionId}/presence-configuration?user_id=${encodeURIComponent(userId)}`);
}
export function savePresenceConfiguration(companionId: string, userId: string, input: PresenceConfigurationInput) {
  return api.put<PresenceConfiguration>(`/companions/${companionId}/presence-configuration?user_id=${encodeURIComponent(userId)}`, input);
}
export function getPresenceSchedule(companionId: string, userId: string) {
  return api.get<PresenceSchedule | null>(`/presence/schedules/${companionId}?user_id=${encodeURIComponent(userId)}`);
}
export function savePresenceSchedule(companionId: string, userId: string, input: PresenceScheduleInput) {
  return api.put<PresenceSchedule>(`/presence/schedules/${companionId}?user_id=${encodeURIComponent(userId)}`, input);
}
export function triggerPresenceSchedule(companionId: string, userId: string, expectedRevision: number) {
  return api.post<PresenceOccurrence>(`/presence/schedules/${companionId}/trigger?user_id=${encodeURIComponent(userId)}`, { expected_revision: expectedRevision });
}
export function listPresenceOccurrences(companionId: string, userId: string, limit = 12) {
  return api.get<PresenceOccurrence[]>(`/presence/schedules/${companionId}/occurrences?user_id=${encodeURIComponent(userId)}&limit=${limit}`);
}
